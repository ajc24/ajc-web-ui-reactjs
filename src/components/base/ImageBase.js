/**
 * Developed by Anthony Cox in 2024
 */
import React from 'react';
import PropTypes from 'prop-types';

/**
 * Baseline component for rendering images in the application.
 * This component allows for images to be fully customised in a variety of ways.
 */
class ImageBase extends React.Component {
  /**
   * Initialise the image base component
   * @param {any} props 
   */
  constructor(props) {
    super(props);
    this.mouseDownDisableAllEvents = this.mouseDownDisableAllEvents.bind(this);
    this.onLoadSetImageOpacityToVisible = this.onLoadSetImageOpacityToVisible.bind(this);

    /* Create the references for this component */
    this.imageRef = React.createRef();
  }

  componentDidMount() {
    /* Set the functionality to be executed when the on load event is fired */
    this.imageRef.current.addEventListener('load', this.onLoadSetImageOpacityToVisible);

    /* Set the functionality to be executed when the on mouse down event is fired */
    if (this.props.onClick !== undefined) {
      this.imageRef.current.addEventListener('mousedown', this.props.onClick);
    } else {
      this.imageRef.current.addEventListener('mousedown', this.mouseDownDisableAllEvents);
    }
  }

  /**
   * Disables all mouse down events on the image
   * @param {Event} event 
   */
  mouseDownDisableAllEvents(event) {
    event.preventDefault();
  }

  /**
   * Sets the opacity of the image to visible when the on load event is fired.
   * This function is especially useful when used in conjunction with transitions to fade loaded images into view.
   */
  onLoadSetImageOpacityToVisible() {
    this.imageRef.current.style.opacity = 1;
    this.imageRef.current.removeEventListener('load', this.onLoadSetImageOpacityToVisible);
  }

  render() {
    /* Set the alternative text for the image - assume a decorative image and only change if it is an informative or functional image */
    let altText = '';
    if (this.props.alt !== undefined && typeof this.props.alt === 'string' && this.props.alt.length > 0) {
      altText = `${this.props.alt}`;
    }
    /* Set the size for the image - assume a default size or set the custom size if provided */
    let imageHeight = 100;
    let imageWidth = 100;
    if (this.props.height !== undefined && typeof this.props.height === 'number') {
      /* Set the custom height for the image */
      imageHeight = this.props.height;
    }
    if (this.props.width !== undefined && typeof this.props.width === 'number') {
      /* Set the custom width for the image */
      imageWidth = this.props.width;
    }
    return (
      <React.Fragment>
        <img id={this.props.id} src={this.props.src} width={imageWidth} height={imageHeight} alt={altText} ref={this.imageRef} />
      </React.Fragment>
    );
  }
}
ImageBase.propTypes = {
  /** The alternate text to be displayed / read out by screen readers. */
  alt: PropTypes.string,
  /** The height of the image. */
  height: PropTypes.number,
  /** The unique identifier for this component. */
  id: PropTypes.string.isRequired,
  /** The custom functionality to be executed when the mouse down on image event is fired. */
  onClick: PropTypes.func,
  /** The image data to be displayed. */
  src: PropTypes.oneOfType([ PropTypes.string, PropTypes.object ]).isRequired,
  /** The width of the image. */
  width: PropTypes.number,
};
export default ImageBase;
