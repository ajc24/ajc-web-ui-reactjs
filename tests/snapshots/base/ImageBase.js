/**
 * Developed by Anthony Cox in 2024
 */
import 'jsdom-global/register';
import React from 'react';
import { TestDev } from 'ajc-testing-code';
import ImageBase from '../../../src/components/base/ImageBase';
import benjiLogo from '../../../src/components/images/files/logo-benji.png';

describe('Image Base', () => {
  describe('Decorative image with default parameters', () => {
    let snapshot;
    
    beforeAll(() => {
      snapshot = TestDev.createSnapshot(
        <React.Fragment>
          <ImageBase id="decorative-image-1" src={benjiLogo} />
        </React.Fragment>
      );
    });

    it('verifies the snapshot for the component', () => {
      expect(snapshot).toMatchSnapshot();
    });
  });

  describe('Informative image with custom parameters', () => {
    let snapshot;
    
    beforeAll(() => {
      snapshot = TestDev.createSnapshot(
        <React.Fragment>
          <ImageBase id="informative-image-1" src={benjiLogo} height={500} width={300} alt="Informative image alt text" />
        </React.Fragment>
      );
    });

    it('verifies the snapshot for the component', () => {
      expect(snapshot).toMatchSnapshot();
    });
  });
});
