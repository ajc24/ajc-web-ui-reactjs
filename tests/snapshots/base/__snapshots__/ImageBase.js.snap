// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Base Decorative image with default parameters verifies the snapshot for the component 1`] = `"<img id="decorative-image-1" class="image" width="100" height="100" alt="" src="[object Object]">"`;

exports[`Image Base Informative image with custom parameters verifies the snapshot for the component 1`] = `"<img id="informative-image-1" class="image" width="300" height="500" alt="Informative image alt text" src="[object Object]">"`;
